### PR TITLE
remove dropdown.js reliance on roles and fix keyboard navigation

### LINKED ISSUE #21159: Can't open Dropdown via space bar in Firefox
In v4, using Firefox.  Hitting spacebar to open the dropdown menu, it opens and quickly closes

steps to reproduce:
- focused on the dropdown components by tab
- press "spacebar" one time
- the menu is display, but then hide again.

Expected behavior: menu could be shown when press spacebar
Actual result: the menu does not open correct. Actually it opens and close quickly

#20303 v3.

### LINKED ISSUE #21941: Dropdown closing on mousedown instead of mouseup
I noticed in the last release (v4.0.0-alpha.6) a new dropdown behavior was introduced, my question is: is this new behavior intentional? because the new interaction seems to be a little odd compared with the old one (closing dropdowns on mouseup).

Testing on Google Chrome Version 55.0.2883.87 (64-bit), Linux Ubuntu Gnome

Thanks in advance

### PR DESCRIPTION
**Pull request by pvdlg**
Per migration documentation in https://github.com/twbs/bootstrap/commit/4d8d8bdab4b0e2931fcf92ce30d772b05a989434:

> - Rebuilt dropdown styles and markup to provide easy, built-in support for `<a>` and `<button>` based dropdown items.
> - Dropdown items now require `.dropdown-item`.

In addition fix the the space key behavior on firefox for `<button>` elements: #21159. Thanks to @RyanThomasMusser for the analysis [here](https://github.com/twbs/bootstrap/issues/20303#issuecomment-234697222). Fixes #21159.

This PR:
- Remove the dependency to `role="menu"` or `role="listbox"` and rely on the class `dropdown-menu` instead
- Remove the dependency to the `a` element for the dropdown-item and rely on the class `dropdown-item` instead
- Remove the necessity to wrap dropdown-item in `li` elements

With this modification the javascript now matches the markup in the documentation.
The keyboard navigation works again now in [dropdowns](http://v4-alpha.getbootstrap.com/components/dropdowns/).

Also fixes #21941 (see [comment](https://github.com/twbs/bootstrap/issues/21941#issuecomment-278502630))

### CODE REVIEW

**Comment by vanduynslagerp** [comment]
Update unit test with new markup.
Per [comment](https://github.com/twbs/bootstrap/pull/21375#issuecomment-269697431)

**Comment by jump020305** [comment]
@all can you help me with this one http://stackoverflow.com/questions/41645884/accessibility-for-select-fields-in-ie11

**Comment by vanduynslagerp** [comment]
@bardiharborow I see that you added a lot of commit on this branch from v4-dev.
Now the branch contains modifications to the dist file and many other that were not in my PR. Also it fails the travis build. Do you know what happened here ?

**Comment by mdo** [comment]
This PR is a little wonky right now, but any chance for a review from @Johann-S and @bardiharborow on the core changes at https://github.com/twbs/bootstrap/pull/21535/files#diff-bfe9dc603f82b0c51ba7430c1fe4c558?

**Comment by vanduynslagerp** [comment]
I merged the last changes from v4-dev and for some reason the diff in the PR show all these changes even though it shouldn't as they are not different from the last v4-dev.
Looking at the diff between v4-dev and my dropdown-keyboard branch show the proper diff: https://github.com/twbs/bootstrap/compare/v4-dev...vanduynslagerp:dropdown-keyboard

**Comment by Johann-S** [comment]
It's a very good work thank you @vanduynslagerp 👍 

IMO we shouldn't remove dependency to ``role="menu"`` or ``role="listbox"`` for the accessibility but I'm not an expert in this subject we should call the ARIA guy @patrickhlauke 😄

**Comment by vanduynslagerp** [comment]
@Johann-S the PR remove the dependency to `role="menu"` and `role="listbox"` in the JS selectors so it's consistent with other JS module and rely on classes for selector.
Like other components in Bootstrap, the use of `role="menu"`, `role="listbox"` or `aria-` for accessibility is recommended in the documentation and examples.

File: js/src/dropdown.js [hunk 1]
```diff
@@ -25,10 +25,11 @@ const Dropdown = (($) => {
   const JQUERY_NO_CONFLICT       = $.fn[NAME]
   const ESCAPE_KEYCODE           = 27 // KeyboardEvent.which value for Escape (Esc) key
   const SPACE_KEYCODE            = 32 // KeyboardEvent.which value for space key
+  const TAB_KEYCODE              = 9 // KeyboardEvent.which value for tab key
   const ARROW_UP_KEYCODE         = 38 // KeyboardEvent.which value for up arrow key
   const ARROW_DOWN_KEYCODE       = 40 // KeyboardEvent.which value for down arrow key
   const RIGHT_MOUSE_BUTTON_WHICH = 3 // MouseEvent.which value for the right button (assuming a right-handed mouse)
-  const REGEXP_KEYDOWN = new RegExp(`${ARROW_UP_KEYCODE}|${ARROW_DOWN_KEYCODE}|${ESCAPE_KEYCODE}|${SPACE_KEYCODE}`)
+  const REGEXP_KEYDOWN = new RegExp(`${ARROW_UP_KEYCODE}|${ARROW_DOWN_KEYCODE}|${ESCAPE_KEYCODE}`)
```
Comment (Johann-S) [inline js/src/dropdown.js]: Can you change the indent please ?

**Review by Johann-S — CHANGES_REQUESTED** [review]

**Review by Johann-S — APPROVED** [review]

**Comment by Johann-S** [comment]
Sorry but your branch is out-of-date so can you update ?

**Comment by Johann-S** [comment]
Again but I arrived too late and your branch isn't up to date and #21802 too

**Comment by Johann-S** [comment]
Can you notice me when you update your branch ? Or can you allow me to update your branch ? Thank you

**Comment by vanduynslagerp** [comment]
I gave you permission

**Review by patrickhlauke — APPROVED** [review]
Looks good, and does indeed fix the issue of Firefox/space. Good stuff

**Pull request merged**